File: solution/027RemoveElement/027RemoveElement.go
package main

import "fmt"

func removeElement(nums []int, val int) int {
	i := len(nums) - 1
	j := len(nums) - 1
	for j >= 0 {
		if nums[j] == val {
			nums[j], nums[i] = nums[i], nums[j]
			i--
		}
		j--
	}
	return i + 1
}

func main() {
	nums := []int{3, 2, 2, 3, 1}
	fmt.Printf("%v\n", nums[:removeElement(nums, 3)])
}

File: src/027RemoveElement/027RemoveElement.rs
impl Solution {
    pub fn remove_element(nums: &mut Vec<i32>, val: i32) -> i32 {
        let n = nums.len();
        if n == 0 {return 0;}
        let mut i:i32 = -1;
        let mut j:usize = 0;
        while j < n {
            while j < n && nums[j] == val {
                j += 1;
            }
            if j < n {
                i += 1;
                nums.swap(i as usize, j);
                j += 1;
            }
        }
        i+1
    }
}
